### ISSUE TITLE
optimize_for_inference output broken graph

### ISSUE THREAD

**Issue by YijinLiu** [tensorflow/tensorflow#21566]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Ubuntu 18.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
No.
- **TensorFlow installed from (source or binary)**:
Source.
- **TensorFlow version (use command below)**:
1.10.0
- **Python version**:
2.7
- **Bazel version (if compiling from source)**:
0.15.2
- **GCC/Compiler version (if compiling from source)**:
7.3.0
- **CUDA/cuDNN version**:
9.2/7.1.4.18
- **GPU model and memory**:
GTX 960 / 4G
- **Exact command to reproduce**:
I trained a ssdlite_mobilenet_v2 model and run the following command to optimize the frozen graph.

python -m tensorflow.python.tools.optimize_for_inference \
    --input=frozen_inference_graph.pb \
    --output=optimized_frozen_inference_graph.pb  \
    --input_names=image_tensor --placeholder_type_enum=4 \
    --output_names=num_detections,detection_classes,detection_scores,detection_boxes

It gave a lot of the following warnings but succeeded:
WARNING:tensorflow:Didn't find expected Conv2D input to 'FeatureExtractor/MobilenetV2/expanded_conv/depthwise/BatchNorm/FusedBatchNorm'

However, the outputted graph is broken:
benchmark_model --graph=optimized_frozen_inference_graph.pb --show_flops --input_layer=image_tensor --input_layer_type=uint8 --input_layer_shape=1,300,300,3 --output_layer=detection_boxes,detection_scores,detection_classes,num_detections

2018-08-13 08:05:27.077759: E tensorflow/core/framework/types.cc:102] Unrecognized DataType enum value 161403164
2018-08-13 08:05:27.080527: E tensorflow/tools/benchmark/benchmark_model.cc:275] Could not create TensorFlow Session: Invalid argument: Input 0 of node Preprocessor/map/while/add/y was passed int32 from Preprocessor/map/while/Switch:1 incompatible with expected unknown dtype enum (161403164)_ref.

**Comment by JianboTang** [comment]
I encounter the same problem ...

**Comment by zh794390558** [comment]
same problem using quantize_graph or transformer_graph for quantize, and run on x86. tf version 1.9

**Comment by YijinLiu** [comment]
I just tested the latest 1.11.0-rc2 release. The problem is still there. Is anything going to take a look at this?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by petewarden** [comment]
Unfortunately optimize_for_inference has become outdated over the last couple of years, and needs to be deprecated. We have an alternative in the Graph Transform Tool (though this is also getting a bit out of date): https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/graph_transforms/README.md#optimizing-for-deployment

**Comment by ema987** [comment]
Hi @petewarden, shall we use then Graph Transform Tool in place of optimize_for_inference wherever it is used? Is there a newer alternative since you say it's becoming out of date? Thank you.

**Comment by petewarden** [comment]
You're correct, I do suggest the GTT in place of this script, since it should be more reliable.
